File: scripts/run_conformance.py
"""Run the canonical Python question-quality conformance fixtures."""

from __future__ import annotations

import argparse
import json
import pathlib
import sys

sys.path.insert(0, str(pathlib.Path(__file__).resolve().parent))
from question_quality import conformance_result, equivalent_strings, parse_value, validate_question


ROOT = pathlib.Path(__file__).resolve().parents[1]
FIXTURES = ROOT / "scripts" / "spec" / "conformance-fixtures.json"


def assert_fixtures(fixtures: dict, result: dict) -> None:
    for case, actual in zip(fixtures["parse"], result["parse"]):
        expected = case["expect"]
        if expected == "PARSE_FAILED":
            if actual["result"].get("status") != "PARSE_FAILED" or actual["result"].get("code") != case.get("code"):
                raise AssertionError(f"parse fixture failed: {case['input']}: {actual}")
        elif actual["result"] != expected:
            raise AssertionError(f"parse fixture failed: {case['input']}: {actual['result']} != {expected}")
    for case, actual in zip(fixtures["equivalence"], result["equivalence"]):
        if actual["equal"] != case["equal"] or (not case["equal"] and actual["code"] != case.get("code")):
            raise AssertionError(f"equivalence fixture failed: {case}: {actual}")
    for case, actual in zip(fixtures["forbidden_fixes"], result["forbidden_fixes"]):
        if actual["verdict"] != "ERROR" or actual["code"] != case["code"]:
            raise AssertionError(f"forbidden fixer fixture failed: {case}: {actual}")
    for case, actual in zip(fixtures["questions"], result["questions"]):
        expected = case["expect"]
        if actual["verdict"] != expected["verdict"] or actual["codes"] != expected["codes"]:
            raise AssertionError(f"question fixture failed: {case['id']}: {actual}")
    for case, actual in zip(fixtures["bounded_termination"], result["bounded_termination"]):
        if actual["code"] != case["expect"]["code"] or not actual["must_terminate"]:
            raise AssertionError(f"termination fixture failed: {case['id']}: {actual}")


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--output", default=str(ROOT / "data" / "conformance" / "python.json"))
    args = parser.parse_args()
    fixtures = json.loads(FIXTURES.read_text(encoding="utf-8"))
    result = conformance_result(fixtures)
    assert_fixtures(fixtures, result)
    output = json.dumps(result, ensure_ascii=False, sort_keys=True, separators=(",", ":"))
    output_path = pathlib.Path(args.output)
    output_path.parent.mkdir(parents=True, exist_ok=True)
    with output_path.open("w", encoding="utf-8", newline="\n") as handle:
        handle.write(output + "\n")
    print(f"Python conformance passed: {args.output}")
    return 0


if __name__ == "__main__":
    try:
        raise SystemExit(main())
    except (AssertionError, OSError, json.JSONDecodeError) as exc:
        print(f"Conformance failed: {exc}", file=sys.stderr)
        raise SystemExit(1)

File: scripts/run-conformance.js
'use strict';

const fs = require('fs');
const path = require('path');
const {
  conformanceResult,
} = require('./question-quality');

const root = path.resolve(__dirname, '..');
const fixturesPath = path.join(root, 'scripts', 'spec', 'conformance-fixtures.json');
const outputIndex = process.argv.indexOf('--output');

const fixtures = JSON.parse(fs.readFileSync(fixturesPath, 'utf8'));
const result = conformanceResult(fixtures);
const fail = message => { throw new Error(message); };

fixtures.parse.forEach((item, index) => {
  const actual = result.parse[index].result;
  if (item.expect === 'PARSE_FAILED') {
    if (actual.status !== 'PARSE_FAILED' || actual.code !== item.code) fail(`parse fixture failed: ${item.input}`);
  } else if (JSON.stringify(actual) !== JSON.stringify(item.expect)) {
    fail(`parse fixture failed: ${item.input}: ${JSON.stringify(actual)}`);
  }
});
fixtures.equivalence.forEach((item, index) => {
  const actual = result.equivalence[index];
  if (actual.equal !== item.equal || (!item.equal && actual.code !== item.code)) fail(`equivalence fixture failed: ${item.a} / ${item.b}`);
});
fixtures.forbidden_fixes.forEach((item, index) => {
  const actual = result.forbidden_fixes[index];
  if (actual.verdict !== 'ERROR' || actual.code !== item.code) fail(`forbidden fixer fixture failed: ${item.file}`);
});
fixtures.questions.forEach((item, index) => {
  const actual = result.questions[index];
  if (actual.verdict !== item.expect.verdict || JSON.stringify(actual.codes) !== JSON.stringify(item.expect.codes)) fail(`question fixture failed: ${item.id}`);
});
fixtures.bounded_termination.forEach((item, index) => {
  const actual = result.bounded_termination[index];
  if (actual.code !== item.expect.code || !actual.must_terminate) fail(`termination fixture failed: ${item.id}`);
});

function sortKeys(value) {
  if (Array.isArray(value)) return value.map(sortKeys);
  if (value && typeof value === 'object') {
    return Object.keys(value).sort().reduce((out, key) => {
      out[key] = sortKeys(value[key]);
      return out;
    }, {});
  }
  return value;
}

const output = JSON.stringify(sortKeys(result), null, 0);
const outputPath = outputIndex >= 0 && process.argv[outputIndex + 1]
  ? process.argv[outputIndex + 1]
  : path.join(root, 'data', 'conformance', 'javascript.json');
fs.mkdirSync(path.dirname(outputPath), { recursive: true });
fs.writeFileSync(outputPath, `${output}\n`, 'utf8');
console.log(`JavaScript conformance passed: ${outputPath}`);
